File: qdblog/renderer.py
import os

from jinja2 import Environment, FileSystemLoader
from pyquery import PyQuery as pq


env = Environment(loader=FileSystemLoader('./templates'))
post_cache = {}


def render_single(file, force=False):
    post = pq(read_post(file))
    if len(post(('article.draft'))) and not force:
        return False
    post_dict = {
        'title': post('h1').text(),
        'html': read_post(file)
    }
    print post_dict
    template = env.get_template('single_post.html')
    output = template.render({ 'post': post_dict })
    write_post(file, unicode(output).encode('utf-8'))
    return True


def read_post(file):
    if not post_cache.get(file):
        post_path = './posts/%s' % (file)
        post_file = open(post_path, 'r')
        post_cache[file] = post_file.read()
    return post_cache.get(file)


def write_post(file, content):
    post_path = './rendered/%s' % (file)
    post_file = open(post_path, 'w')
    print post_path
    post_file.write(content)


def render_index(num_posts=5):
    path = './posts'
    posts = []
    posts_dict = []
    for file in os.listdir(path):
        post = pq(read_post(file))
        date = post('.meta .posted').text()
        posts.append((file, date))
    print posts
    posts = sorted(posts, cmp=lambda a, b: cmp(b[1], a[1]))[:10]
    for post, ctime in posts:
        doc = pq(read_post(post))
        if not len(doc(('article.draft'))):
            doc('.meta').append(' <a href="/blog/%s">#</a>' % (post))
            posts_dict.append({ 'html': doc.outerHtml() })
    template = env.get_template('index.html')
    output = template.render({ 'posts': posts_dict })
    write_post('index.html', unicode(output).encode('utf-8'))


def render_list(files, force=False):
    post_cache = {}
    files_rendered = 0;
    for filename in files:
        print 'rendering %s... ' %  (filename),
        if render_single(filename, force):
            print 'done'
            files_rendered += 1
        else:
            print 'skipped'

    print '%d file%s rendered' % (files_rendered, '' if files_rendered == 1 else 's')

File: renderer.js
var fs = require('fs');
var jsdom = require('jsdom').jsdom;
var nunjucks = require('nunjucks');

var post_cache = {};

var env = new nunjucks.Environment(new nunjucks.FileSystemLoader('templates'));

var postDir = './posts/';

var NUM_INDEX_POSTS = 3;

function read_post(path, force) {
    if (path in post_cache && !force) {
        return post_cache[path];
    }
    var contents = fs.readFileSync(postDir+path, 'utf-8');
    post_cache[path] = contents;
    return contents;
}

function write_post(path, content) {
    fs.writeFileSync('rendered/'+path, content, 'utf-8');
}

function singlePost(post) {
    var tmpl = env.getTemplate('single_post.html');
    var post_markup = read_post(post)

    var doc = jsdom(post_markup);
    var window = doc.createWindow();

    var article = window.document.getElementsByTagName('article')[0];

    var p = {
        title: extractTitle(window),
        date: article.getAttribute('data-posted'),
        html: article.innerHTML
    };

    return tmpl.render({ post: p });
}

function renderAll() {
    var posts = fs.readdirSync(postDir);
    var out = '';
    posts.forEach(function (post) {
        var rendered = singlePost(post);
        write_post(post, rendered);
        out += 'wrote ' + post + '\n';
        // console.log(fs.statSync(dir+'/'+post));
    });
    out += renderIndex();
    return out;
}

function renderIndex() {
    var posts = fs.readdirSync(postDir);
    var out = 'rendering index...\n';
    var tuples = posts.map(function(post) {
        var doc = jsdom(read_post(post));
        var window = doc.createWindow();
        var article = window.document.getElementsByTagName('article')[0];
        if (article.getAttribute('class').indexOf('draft')+1) return;
        var date = article.getAttribute('data-posted');
        return [date, post, article];
    });
    tuples.sort(function(a, b) {
        return (a[0] < b[0]) ? 1 : -1;
    });
    var p = [];
    for (var i=0; i<NUM_INDEX_POSTS; i++) {
        tuple = tuples[i];
        console.log(tuple[2].innerHTML);
        p.push({
            html: tuple[2].innerHTML,
            date: tuple[0],
            permalink: '/blog/' + tuple[1]
        });
        out += tuple[1] + '\n';
    }
    var tmpl = env.getTemplate('index.html');
    write_post('index.html', tmpl.render({ posts: p }));
    return out;
}

module.exports = {
    all: renderAll,
    rank: renderIndex,
    single_post: singlePost
};

function extractTitle(win) {
    var h1s = win.document.getElementsByTagName('h1');
    if (!h1s.length) return '';
    return text(h1s[0]);
}

function text(node) {
    var cn = node.childNodes;
    var t='';

    if (cn.length) {
        for (var i=0; i<cn.length; i++) {
            t += text(cn[i]);
        }
    } else {
        t = node.textContent;
    }

    return t;
}
